docs/connections.md

```diff
@@ -0,0 +1,597 @@
+however lambda body is THE scope where argument is introduced and is same as
+lambda code block, it should be an error. 
+Or should it get some special rule?
+
+---
+
+### Function definitions
+If the assignment's left-hand side is a prefix application chain, where the
+left-most element (the callable) is a variable name, the assignment is said to
+be a function definition. Each prefix argument is converted into a lambda
+argument in the assignment body.
+
+```
+«log_name» object = »print« object.»name«
+```
+is desugared into:
+```
+«log_name» = object -> »print« object.»name«
+```
+which in turn can be desugared into:
+```
+«log_name» = object -> »print« (»name« object)
+```
+
+This desugaring shows why only `log_name` is introduced into the scope, while
+`object` is visible only in the definition's body.
+
+If the operator appears in the function name position, it can be defined as
+well:
+```
+«^» a n = a * a ^ (n - 1)
+```
+
+This introduces name `^` into the scope. It uses already defined `*` and `-`
+operators. (to avoid clutter the operators are not marked with »«)
+
+### Pattern matching
+If the assignment's left-hand side is a prefix application chain where the
+function is a constructor name, it will be desugared into a pattern match.
+
+Example:
+```
+»Some« «value» = »get_opt«
+```
+
+will be desugared into:
+
+```
+«value» = case »get_opt« of
+  »Some« b -> b
+  _      -> error 
+```
+
+Therefore, only `value` will be introduced into the parent scope. `Some` and
+`get_opt` must be defined, the former being an atom with at least single field.
+
+Using operators in the infix position will also attempt to pattern match its
+operands. For example:
+```
+«x»,«y» = »get_position«
+```
+
+TODO Describe what it gets desugared to when there is more than one variable
+name introduced? Now it is inconsistent, if we say that only canonical form of
+assignment is `variable_name = value`.
+
+### Extensions methods
+If the application target uses accessor operator `.`, e.g. `Int.add`, the last
+segment of target is the introduced identifier and the previous segments are
+used to type the implicit `this` parameter. 
+
+For example:
+```
+»Foo«.«bar» = 5
+```
+translated to:
+```
+«bar» this:»Foo« = 5
+```
+
+Which is then desugared into a lambda. The introduced name is only `bar`.
+
+If there are any prefix application arguments following the accessor-style
+target, they will be treated as arguments following implicit `this`.
+
+### Overloading
+Only the methods that take `this` as the first parameter can be overloaded. Each
+overload of the given name must have different type of `this`.
+
+However, the type of `this` will be often inferred by the typechecker and it
+IDE cannot tell if given overloads are valid or not.
+
+Example:
+
+```
+«foo» this:«a» = »body1«
+
+«foo» this:«b» = »body2«
+```
+
+In this case `a` and `b` for each `foo` definition will be inferred by the
```

Comment:
This is true, but not for a while.

Reply:
Doing this requires a typechecker.

I should probably note here that we _do_, however currently support overloading of a method on a given type as long as the arity is different, so you can define

```
Foo.bar = x -> x + 1
Foo.bar = x y -> x + 1
```